### PR TITLE
Add a SpanDataset

### LINKED ISSUE #976: SpanDataset would simplify splitting up pre-merged datasets
### Feature Description
Not all datasets can be easily split ahead of time, a SpanDataset would allow splitting up
datasets after the fact by specifying an offset and bound for the span.

#### Use Case
Splitting out train, test and validation sets is quite common but not all initial
datasets have such splits baked in.

#### Additional Context
Example API:

```
std::vector<Tensor> tensormap = {fl::rand({100, 200, 300}), fl::rand({150, 300})};
auto tensords = std::make_shared<TensorDataset>(tensormap);

SpanDataset frontspands(tensords, 0, 13);
SpanDataset backspands(tensords, 13);

// Check `size` method
ASSERT_EQ(frontspands.size(), 13);
ASSERT_EQ(backspands.size(), 287);

// Values using `get` method
auto ff1 = frontspands.get(10);
ASSERT_EQ(ff1.size(), 2);
ASSERT_TRUE(allClose(ff1[0], tensormap[0](fl::span, fl::span, 10)));
ASSERT_TRUE(allClose(ff1[1], tensormap[1](fl::span, 10)));
auto ff2 = backspands.get(10);
ASSERT_EQ(ff2.size(), 2);
ASSERT_TRUE(allClose(ff2[0], tensormap[0](fl::span, fl::span, 13 + 10)));
ASSERT_TRUE(allClose(ff2[1], tensormap[1](fl::span, 13 + 10)));
```

### PR DESCRIPTION
**Pull request by bwasti**
Add a SpanDataset to simplify splitting training and validation data from merged initial datasets

**Original Issue**: closes #976

### Summary
As explained in the issue, this feature simplifies splitting up datasets into test, train and
validation sets.

### Test Plan (required)
See additions to `flashlight/fl/test/dataset/DatasetTest.cpp`

### CODE REVIEW

File: flashlight/fl/dataset/SpanDataset.h [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ * A view into an underlying dataset with an offset and optional bounded length.
+ *
+ * The size of the `SpanDataset` is either specified for the size of the input dataset
+ * accounting for the offset.
+ *
+ * We have, for example `SpanDataset(ds, 13).get(i) == ds.get(13 + i)`
+ *
+ * Example:
+  \code{.cpp}
+  // Make a datasets
```
Comment (jacobkahn) [inline flashlight/fl/dataset/SpanDataset.h:26]:
```suggestion
  // Make a dataset
```

File: flashlight/fl/dataset/SpanDataset.cpp [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  if (size_ + offset_ > dataset_->size()) {
+    throw std::out_of_range(
+        "Dataset length out of range (larger than underlying dataset)");
+  }
+}
+
+std::vector<Tensor> SpanDataset::get(const int64_t idx) const {
+  checkIndexBounds(idx);
+
+  std::vector<Tensor> result;
```
Comment (jacobkahn) [inline flashlight/fl/dataset/SpanDataset.cpp:28]: small optimization: we can reserve or preemptively size `result` with `f.size()`

File: flashlight/fl/dataset/SpanDataset.h [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  \endcode
+ */
+class SpanDataset : public Dataset {
+ public:
+  /**
+   * Creates a SpanDataset.
+   * @param[in] dataset The underlying dataset.
+   * @param[in] offset The starting index of the new dataset relative to the
+   * underlying dataset.
+   * @param[in] length The size of the new dataset (if -1, uses previous size
```
Comment (jacobkahn) [inline flashlight/fl/dataset/SpanDataset.h:50]: `previous size` -- can you clarify this to be the input dataset size?

File: flashlight/fl/dataset/datasets.h [hunk 4]
```diff
@@ -17,6 +17,7 @@
 #include "flashlight/fl/dataset/MergeDataset.h"
 #include "flashlight/fl/dataset/PrefetchDataset.h"
 #include "flashlight/fl/dataset/ResampleDataset.h"
+#include "flashlight/fl/dataset/SpanDataset.h"
```
Comment (jacobkahn) [inline flashlight/fl/dataset/datasets.h:20]: clang format to alphabetize?

File: flashlight/fl/dataset/SpanDataset.h [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+ * accounting for the offset.
+ *
+ * We have, for example `SpanDataset(ds, 13).get(i) == ds.get(13 + i)`
+ *
+ * Example:
+  \code{.cpp}
+  // Make a datasets
+  auto makeDataset = []() {
+    auto tensor = fl::rand({5, 4, 10});
+    std::vector<Tensor> fields{tensor};
```
Comment (jacobkahn) [inline flashlight/fl/dataset/SpanDataset.h:29]: nit -- do you really need `makeDataset`?

File: flashlight/fl/dataset/SpanDataset.h [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+#include "flashlight/fl/dataset/Dataset.h"
+
+#include <vector>
+
+namespace fl {
+
+/**
+ * A view into an underlying dataset with an offset and optional bounded length.
+ *
+ * The size of the `SpanDataset` is either specified for the size of the input dataset
```
Comment (jacobkahn) [inline flashlight/fl/dataset/SpanDataset.h:19]: nit
```suggestion
 * The size of the `SpanDataset` can be explicitly specified but defaults to the size of the input dataset
```

File: flashlight/fl/dataset/SpanDataset.h [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+   * @param[in] dataset The underlying dataset.
+   * @param[in] offset The starting index of the new dataset relative to the
+   * underlying dataset.
+   * @param[in] length The size of the new dataset (if -1, uses previous size
+   * minus the offset)
+   */
+  explicit SpanDataset(
+      std::shared_ptr<const Dataset> dataset,
+      const int64_t offset,
+      const int64_t length = -1);
```
Comment (jacobkahn) [inline flashlight/fl/dataset/SpanDataset.h:56]: nit -- maybe rename length `size`? you use that in the docs/I think that's clearer

**Review by jacobkahn — CHANGES_REQUESTED** [review]
Thanks for this — super useful little abstraction. A few quick things:

File: flashlight/fl/dataset/SpanDataset.cpp (hunk 2 above)
Comment (bwasti) [reply flashlight/fl/dataset/SpanDataset.cpp:28]: oh can I just return `dataset_->get(idx + offset_);`?

File: flashlight/fl/dataset/SpanDataset.h (hunk 5 above)
Comment (bwasti) [reply flashlight/fl/dataset/SpanDataset.h:29]: nope

File: flashlight/fl/dataset/datasets.h (hunk 4 above)
Comment (bwasti) [reply flashlight/fl/dataset/datasets.h:20]: huh, I ran clang-format. is that captured in the repo's .clang-format file?

Comment (jacobkahn) [reply flashlight/fl/dataset/datasets.h]: should be, but if you can it, dw about it

**Pull request closed without merging**